Helm implementation based on cli wrapper

LucasBoisserie opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
New Helm Module implementation based on cli wrapper like terraform module
check #61546 for why a new module
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- New Module Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->

##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->

yelhouti commented:
@LucasBoisserie thanks a lot, I'm using it locally, and I love it

stanislaw55 commented:
Is it scheduled for test/review/merge? It would be very very nice to have Ansible helm module which support Helm 3.

benjamink commented:
This would be very useful for me as well.  Any updates?

MarkusTeufelberger commented on lib/ansible/modules/cloud/misc/helm_cli.py:
```diff
@@ -0,0 +1,532 @@
+
+
+# Get chart info
+def fecth_chart_info(command, chart_ref):
```
```suggestion
def fetch_chart_info(command, chart_ref):
```

LucasBoisserie replied:
Thanks for the suggestion

MarkusTeufelberger commented on lib/ansible/modules/cloud/misc/helm_cli.py:
```diff
@@ -0,0 +1,532 @@
+"""
+
+try:
+    import yaml
```
Is that guaranteed to be available everywhere or should it be a requirement listed with the module?

LucasBoisserie replied:
import yaml is guaranteed by the pyaml module, one of ansible depency

MarkusTeufelberger replied:
Yes, but this module doesn't run on machines where ansible is installed...

LucasBoisserie replied:
Interesting !
Can you give me more information on your setup ?
I test with ansible 2.7 and 2.8 on clean virtualenv and i can't reproduce your issue
```
$ pip list
Package      Version
------------ -------
ansible      2.8.5
asn1crypto   1.1.0
cffi         1.12.3
cryptography 2.7
Jinja2       2.10.3
MarkupSafe   1.1.1
pip          19.2.3
pycparser    2.19
PyYAML       5.1.2
setuptools   41.2.0
six          1.12.0
wheel        0.33.6
```

MarkusTeufelberger replied:
This is a module, it gets run on remote hosts, not your Ansible host. If you install Ansible on machines you manage, you're doing something wrong.

Maybe you'll see what I mean once you add the integration tests? As far as I know these are supposed to be nearly mandatory for new modules anyways: https://docs.ansible.com/ansible/latest/dev_guide/developing_modules_checklist.html

LucasBoisserie replied:
I see what you mean !
Don't worry i didn't install ansible on my managed machine ;) 
I launch the tasks on my localhost because my kubernetes cluster is accessible from it
But i forgot this use case !
I'm agree with you that the yaml import must be in the requirement list !
Thanks 

I will check for tests also !

LucasBoisserie replied:
Integration tests have been added

T1loc commented:
Hello,
Is it possible to use a lookup template for the field values ? 
Like this :
```
---
- helm_cli:
     name: "test"
     values: "{{ lookup('template', somefile.j2 ) }}
```

LucasBoisserie committed:
begin helm 3

remove import by mistake

add mdi to contributor

update helm 3 vars

clean comment

MarkusTeufelberger commented on lib/ansible/modules/cloud/misc/helm_cli.py:
```diff
@@ -0,0 +1,542 @@
+requirements:
+  - "helm"
+  - "yaml"
+  - "tempfile"
```
Isn't `tempfile` part of the Python standard library? Or do you refer to something else?

LucasBoisserie replied:
Yes, exactly !
I removed it

MarkusTeufelberger commented on lib/ansible/modules/cloud/misc/helm_cli.py:
```diff
@@ -0,0 +1,542 @@
+author:
+  - Lucas Boisserie (@LucasBoisserie)
+notes:
+  - Support for helm3 is not finished waiting fix on helm 3
```
I'm not sure what this is supposed to mean. "Waiting fix on helm3"? Which fix?

stanislaw55 replied:
Since stable Helm v3 is coming soon (it hit rc 3 recently and new pre-releases are being published frequently) is would be good to support also Helm 3 out-of-box.

LucasBoisserie replied:
I'm working on it !
I have another branch where the module works with last release of helm3
I haven't merge it yet because I add some integration tests on both helm2 and helm3

LucasBoisserie replied:
Helm 3 support is now ready and tested

MarkusTeufelberger commented on lib/ansible/modules/cloud/misc/helm_cli.py:
```diff
@@ -0,0 +1,542 @@
+        required_together=[
+            ['chart_repo_username', 'chart_repo_password']
+        ],
+        supports_check_mode=True,
```
I'm not too certain about this - it seems as if it would always return "changed" in check mode?

LucasBoisserie replied:
Not exactly !
In check mode, the module return changed only if
* the chart is not install
* the chart need to be upgraded
* the values changed

LucasBoisserie committed:
Helm implementation based on cli wrapper

missing doc type

add support for helm 3

Fix method name fetch_chart_info

Co-Authored-By: MarkusTeufelberger <mteufelberger@mgit.at>

missing space

add python dependencies in requirement and improve import management

Remove false requirements

Commit:
Fixes:
helm versions detection
deletion of helm chart v2
add kube_context global option

Commit:
Update difference between output on helm2 and helm3
Lowercase all keys by default

erhudy commented:
We'd love to see this merged in time for 2.10. Is there any way in which we can offer support to get this pushed through?

dstarner commented:
+1 on available support

T1loc commented:
We added the kube_context option, and fixes some compatibilities between helm 2 and helm 3.
I'm using the patch modules and it works really well. 
I would like to add the --kubeconfig option too.
I don't know what else is missing (integration tests ?)

stanislaw55 commented:
I can offer testing on Helm 3 and few fixes for Helm 3 support

LucasBoisserie committed:
Add Integration tests

MatthiasLohr commented:
Anything new here?

MatthiasLohr commented:
One finding when testing this: I'm trying to install metallb via this module:
```
    - name: test helm
      helm_cli:
        chart_ref: metallb
        release_name: metallb
        chart_repo_url: https://kubernetes-charts.storage.googleapis.com/
        release_namespace: metallb-system
        release_values:
          address-pools:
            - name: default
              protocol: layer2
              addresses:
                - 10.195.1.169
        wait: yes
```

Result:
```
FAILED! => {"changed": false, "command": "/usr/local/sbin/helm --repo=https://kubernetes-charts.storage.googleapis.com/ upgrade -i --reset-values --wait -f=/tmp/tmpTd7kBl.yml --namespace=metallb-system metallb metallb", "msg": "Failure when executing Helm command. Exited 1.\nstdout: Release \"metallb\" does not exist. Installing it now.\n\nstderr: Error: create: failed to create: namespaces \"metallb-system\" not found\n"}
```

Is it intended to manually create these namespaces?

T1loc commented:
Yes, helm 3 is not creating anymore the namespace. You need to do that first

LucasBoisserie commented:
Support for helm 3 and integration tests are done 

Waiting for review
Ping @TimothyVandenbrande @UnderGreen @cheese @drybjed @flaper87 @helldorado @joshainglis @karmab @machacekondra @mnecas @mwperina

MarkusTeufelberger commented on lib/ansible/modules/cloud/misc/helm_cli.py:
```diff
@@ -0,0 +1,605 @@
+  - Matthieu Diehr (@d-matt)
+requirements:
+  - "helm"
+  - "yaml"
```
Maybe add a link to PyPi or something, one of these is a Python library and the other one an external binary.

LucasBoisserie replied:
Links added
Thanks for the suggestion !

MarkusTeufelberger commented on lib/ansible/modules/cloud/misc/helm_cli.py:
```diff
@@ -0,0 +1,605 @@
+      - Required when release_state is set to present
+    required: false
+    type: str
+    version_added: "2.10"
```
Afaik the `version_added` line is not necessary for new modules

LucasBoisserie replied:
version_added cleaned

MarkusTeufelberger commented on lib/ansible/modules/cloud/misc/helm_cli.py:
```diff
@@ -0,0 +1,605 @@
+    state: absent
+    tiller_namespace: helm
+
+# With Helm3
```
Since helm3 doesn't create namespaces, maybe add in a task that creates the `monitoring` namespace using the `k8s` module?

LucasBoisserie replied:
Added too !

LucasBoisserie committed:
Add review suggestions

dstarner commented:
Any update on this? Seems like the checks have passed

MatthiasLohr commented:
I'm already using this for my cluster - works like a charm.

SamyCoenen commented:
Also already using this, works great.

yelhouti commented:
@MarkusTeufelberger @stanislaw55 this waiting on you guys, please review it. Thanks :)

MarkusTeufelberger commented:
> @MarkusTeufelberger @stanislaw55 this waiting on you guys, please review it. Thanks :)

No, it is not. This is waiting on sign-offs from people that wrote modules in the "cloud" namespace. See the comment from ansibot above. You can try the `bot_status` command to learn more. Also please read https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md (linked by the bot in every comment it makes) to learn how code gets merged.

I for example can only review/approve changes in the "crypto" namespace, and only because I am a (co-)author of a module there.

yelhouti commented:
@MarkusTeufelberger Sorry my bad, I saw you in the reviewers section, I though it was managed ansibot, I guess I will be pinging the others then :)

yelhouti commented:
@TimothyVandenbrande @UnderGreen @cheese @drybjed @flaper87 @helldorado @joshainglis @karmab @machacekondra @mnecas @mwperina this waiting on you guys, please review it. Thanks :)

MarkusTeufelberger commented:
Yeah, that's github thinking a single comment (that i explicitly marked as comment) still makes me a reviewer.

MatthiasLohr commented:
@MarkusTeufelberger, maybe not a single comment but the list of users under "Reviewers"?
![image](https://user-images.githubusercontent.com/781926/71171068-2964db80-225d-11ea-8309-6f6fa3c62d3e.png)

MarkusTeufelberger commented:
If you write a comment on code instead this general section in the PR, GitHub lists you in that section. Also you can see by the speech bubble that I didn't do a review (which I would need to approve according to GitHub's workflow), I only left comments. Still: This is not relevant in any way to the way code gets merged in this repository. You need core members or people from this namespace to sign off on this.

T1loc commented:
Maybe something need to be improved. When a helm chart has an extra output the module fail. (Eg Prometheus). The helm output says that some hooks are skipped and then the module fail because he is reading the stdout.

dstarner commented:
@T1loc can you please post the output of this behavior from one of your example tests please?

T1loc commented:
> @T1loc can you please post the output of this behavior from one of your example tests please?

```
helm install bitnami/prometheus-operator --namespace monitoring --generate-name                                                  ✔  docker-desktop/monitoring ⎈  12:47:51
manifest_sorter.go:175: info: skipping unknown hook: "crd-install"
manifest_sorter.go:175: info: skipping unknown hook: "crd-install"
manifest_sorter.go:175: info: skipping unknown hook: "crd-install"
manifest_sorter.go:175: info: skipping unknown hook: "crd-install"
manifest_sorter.go:175: info: skipping unknown hook: "crd-install"
NAME: prometheus-operator-1577447271
LAST DEPLOYED: Fri Dec 27 12:47:55 2019
NAMESPACE: monitoring
STATUS: deployed
REVISION: 1
TEST SUITE: None
NOTES:
** Please be patient while the chart is being deployed **

Watch the Prometheus Operator Deployment status using the command:

    kubectl get deploy -w --namespace monitoring -l app.kubernetes.io/name=prometheus-operator-operator,app.kubernetes.io/instance=prometheus-operator-1577447271

Watch the Prometheus StatefulSet status using the command:

    kubectl get sts -w --namespace monitoring -l app.kubernetes.io/name=prometheus-operator-prometheus,app.kubernetes.io/instance=prometheus-operator-1577447271

Prometheus can be accessed via port "9090" on the following DNS name from within your cluster:

    prometheus-operator-157744-prometheus.monitoring.svc.cluster.local

To access Prometheus from outside the cluster execute the following commands:

    echo "Prometheus URL: http://127.0.0.1:9090/"
    kubectl port-forward --namespace monitoring svc/prometheus-operator-157744-prometheus 9090:9090

Watch the Alertmanager StatefulSet status using the command:

    kubectl get sts -w --namespace monitoring -l app.kubernetes.io/name=prometheus-operator-alertmanager,app.kubernetes.io/instance=prometheus-operator-1577447271

Alertmanager can be accessed via port "9093" on the following DNS name from within your cluster:

    prometheus-operator-157744-alertmanager.monitoring.svc.cluster.local

To access Alertmanager from outside the cluster execute the following commands:

    echo "Alertmanager URL: http://127.0.0.1:9093/"
    kubectl port-forward --namespace monitoring svc/prometheus-operator-157744-alertmanager 9093:9093
```

This is the current output when I try to install the prometheus-operator chart.

keiransteele commented:
@T1loc I couldn't reproduce the above issue. I got the bitnami prometheus-operator to install with no errors.
```
TASK [Deploy latest version of Prometheus Operator chart inside monitoring namespace] ****************************************************************************************************************************************************
task path: /home/keiran/Ansible/helm-cli-test.yaml:13
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: keiran
<127.0.0.1> EXEC /bin/sh -c 'echo ~keiran && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /home/keiran/.ansible/tmp/ansible-tmp-1577491937.348863-65627865877027 `" && echo ansible-tmp-1577491937.348863-65627865877027="` echo /home/keiran/.ansible/tmp/ansible-tmp-1577491937.348863-65627865877027 `" ) && sleep 0'
Using module_utils file /usr/lib/python3.7/site-packages/ansible/module_utils/basic.py
Using module_utils file /usr/lib/python3.7/site-packages/ansible/module_utils/parsing/__init__.py
Using module_utils file /usr/lib/python3.7/site-packages/ansible/module_utils/parsing/convert_bool.py
Using module_utils file /usr/lib/python3.7/site-packages/ansible/module_utils/pycompat24.py
Using module_utils file /usr/lib/python3.7/site-packages/ansible/module_utils/common/_collections_compat.py
Using module_utils file /usr/lib/python3.7/site-packages/ansible/module_utils/common/validation.py
Using module_utils file /usr/lib/python3.7/site-packages/ansible/module_utils/common/_utils.py
Using module_utils file /usr/lib/python3.7/site-packages/ansible/module_utils/_text.py
Using module_utils file /usr/lib/python3.7/site-packages/ansible/module_utils/common/__init__.py
Using module_utils file /usr/lib/python3.7/site-packages/ansible/module_utils/six/__init__.py
Using module_utils file /usr/lib/python3.7/site-packages/ansible/module_utils/common/parameters.py
Using module_utils file /usr/lib/python3.7/site-packages/ansible/module_utils/common/text/converters.py
Using module_utils file /usr/lib/python3.7/site-packages/ansible/module_utils/common/_json_compat.py
Using module_utils file /usr/lib/python3.7/site-packages/ansible/module_utils/common/text/__init__.py
Using module_utils file /usr/lib/python3.7/site-packages/ansible/module_utils/common/file.py
Using module_utils file /usr/lib/python3.7/site-packages/ansible/module_utils/common/text/formatters.py
Using module_utils file /usr/lib/python3.7/site-packages/ansible/module_utils/common/sys_info.py
Using module_utils file /usr/lib/python3.7/site-packages/ansible/module_utils/common/process.py
Using module_utils file /usr/lib/python3.7/site-packages/ansible/module_utils/common/collections.py
Using module_utils file /usr/lib/python3.7/site-packages/ansible/module_utils/distro/__init__.py
Using module_utils file /usr/lib/python3.7/site-packages/ansible/module_utils/distro/_distro.py
Using module file /home/keiran/.ansible/plugins/modules/helm_cli.py
<127.0.0.1> PUT /home/keiran/.ansible/tmp/ansible-local-4961lbzfbf7x/tmp9khnlcx9 TO /home/keiran/.ansible/tmp/ansible-tmp-1577491937.348863-65627865877027/AnsiballZ_helm_cli.py
<127.0.0.1> EXEC /bin/sh -c 'chmod u+x /home/keiran/.ansible/tmp/ansible-tmp-1577491937.348863-65627865877027/ /home/keiran/.ansible/tmp/ansible-tmp-1577491937.348863-65627865877027/AnsiballZ_helm_cli.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '/usr/bin/python /home/keiran/.ansible/tmp/ansible-tmp-1577491937.348863-65627865877027/AnsiballZ_helm_cli.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c 'rm -f -r /home/keiran/.ansible/tmp/ansible-tmp-1577491937.348863-65627865877027/ > /dev/null 2>&1 && sleep 0'
changed: [localhost] => {
    "changed": true,
    "command": "/usr/local/bin/helm --repo=https://charts.bitnami.com/bitnami upgrade -i --reset-values --namespace=monitoring prometheus-operator prometheus-operator",
    "invocation": {
        "module_args": {
            "binary_path": null,
            "chart_ref": "prometheus-operator",
            "chart_repo_password": null,
            "chart_repo_url": "https://charts.bitnami.com/bitnami",
            "chart_repo_username": null,
            "chart_version": null,
            "disable_hook": false,
            "force": false,
            "kube_context": null,
            "kubeconfig_path": null,
            "name": "prometheus-operator",
            "purge": true,
            "release_name": "prometheus-operator",
            "release_namespace": "monitoring",
            "release_state": "present",
            "release_values": {},
            "tiller_host": null,
            "tiller_namespace": "kube-system",
            "update_repo_cache": false,
            "wait": false,
            "wait_timeout": null
        }
    },
    "status": {
        "app_version": "0.34.0",
        "chart": "prometheus-operator-0.7.5",
        "name": "prometheus-operator",
        "namespace": "monitoring",
        "revision": "1",
        "status": "deployed",
        "updated": "2019-12-28 13:12:33.13170322 +1300 NZDT",
        "values": {}
    },
    "stderr": "manifest_sorter.go:175: info: skipping unknown hook: \"crd-install\"\nmanifest_sorter.go:175: info: skipping unknown hook: \"crd-install\"\nmanifest_sorter.go:175: info: skipping unknown hook: \"crd-install\"\nmanifest_sorter.go:175: info: skipping unknown hook: \"crd-install\"\nmanifest_sorter.go:175: info: skipping unknown hook: \"crd-install\"\n",
    "stderr_lines": [
        "manifest_sorter.go:175: info: skipping unknown hook: \"crd-install\"",
        "manifest_sorter.go:175: info: skipping unknown hook: \"crd-install\"",
        "manifest_sorter.go:175: info: skipping unknown hook: \"crd-install\"",
        "manifest_sorter.go:175: info: skipping unknown hook: \"crd-install\"",
        "manifest_sorter.go:175: info: skipping unknown hook: \"crd-install\""
    ],
    "stdout": "Release \"prometheus-operator\" does not exist. Installing it now.\nNAME: prometheus-operator\nLAST DEPLOYED: Sat Dec 28 13:12:33 2019\nNAMESPACE: monitoring\nSTATUS: deployed\nREVISION: 1\nTEST SUITE: None\nNOTES:\n** Please be patient while the chart is being deployed **\n\nWatch the Prometheus Operator Deployment status using the command:\n\n    kubectl get deploy -w --namespace monitoring -l app.kubernetes.io/name=prometheus-operator-operator,app.kubernetes.io/instance=prometheus-operator\n\nWatch the Prometheus StatefulSet status using the command:\n\n    kubectl get sts -w --namespace monitoring -l app.kubernetes.io/name=prometheus-operator-prometheus,app.kubernetes.io/instance=prometheus-operator\n\nPrometheus can be accessed via port \"9090\" on the following DNS name from within your cluster:\n\n    prometheus-operator-prometheus.monitoring.svc.cluster.local\n\nTo access Prometheus from outside the cluster execute the following commands:\n\n    echo \"Prometheus URL: http://127.0.0.1:9090/\"\n    kubectl port-forward --namespace monitoring svc/prometheus-operator-prometheus 9090:9090\n\nWatch the Alertmanager StatefulSet status using the command:\n\n    kubectl get sts -w --namespace monitoring -l app.kubernetes.io/name=prometheus-operator-alertmanager,app.kubernetes.io/instance=prometheus-operator\n\nAlertmanager can be accessed via port \"9093\" on the following DNS name from within your cluster:\n\n    prometheus-operator-alertmanager.monitoring.svc.cluster.local\n\nTo access Alertmanager from outside the cluster execute the following commands:\n\n    echo \"Alertmanager URL: http://127.0.0.1:9093/\"\n    kubectl port-forward --namespace monitoring svc/prometheus-operator-alertmanager 9093:9093\n",
    "stdout_lines": [
        "Release \"prometheus-operator\" does not exist. Installing it now.",
        "NAME: prometheus-operator",
        "LAST DEPLOYED: Sat Dec 28 13:12:33 2019",
        "NAMESPACE: monitoring",
        "STATUS: deployed",
        "REVISION: 1",
        "TEST SUITE: None",
        "NOTES:",
        "** Please be patient while the chart is being deployed **",
        "",
        "Watch the Prometheus Operator Deployment status using the command:",
        "",
        "    kubectl get deploy -w --namespace monitoring -l app.kubernetes.io/name=prometheus-operator-operator,app.kubernetes.io/instance=prometheus-operator",
        "",
        "Watch the Prometheus StatefulSet status using the command:",
        "",
        "    kubectl get sts -w --namespace monitoring -l app.kubernetes.io/name=prometheus-operator-prometheus,app.kubernetes.io/instance=prometheus-operator",
        "",
        "Prometheus can be accessed via port \"9090\" on the following DNS name from within your cluster:",
        "",
        "    prometheus-operator-prometheus.monitoring.svc.cluster.local",
        "",
        "To access Prometheus from outside the cluster execute the following commands:",
        "",
        "    echo \"Prometheus URL: http://127.0.0.1:9090/\"",
        "    kubectl port-forward --namespace monitoring svc/prometheus-operator-prometheus 9090:9090",
        "",
        "Watch the Alertmanager StatefulSet status using the command:",
        "",
        "    kubectl get sts -w --namespace monitoring -l app.kubernetes.io/name=prometheus-operator-alertmanager,app.kubernetes.io/instance=prometheus-operator",
        "",
        "Alertmanager can be accessed via port \"9093\" on the following DNS name from within your cluster:",
        "",
        "    prometheus-operator-alertmanager.monitoring.svc.cluster.local",
        "",
        "To access Alertmanager from outside the cluster execute the following commands:",
        "",
        "    echo \"Alertmanager URL: http://127.0.0.1:9093/\"",
        "    kubectl port-forward --namespace monitoring svc/prometheus-operator-alertmanager 9093:9093"
    ]
}
META: ran handlers
META: ran handlers

PLAY RECAP *******************************************************************************************************************************************************************************************************************************
localhost                  : ok=3    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   
```
```yaml
---

- hosts: localhost
  tasks:
  
    - name: Create helm namespace HELM 3 doesn't create it automatically
      k8s:
        api_version: v1
        kind: Namespace
        name: "monitoring"
        wait: true

    - name: Deploy latest version of Prometheus Operator chart inside monitoring namespace
      helm_cli:
        name: prometheus-operator
        chart_ref: prometheus-operator
        chart_repo_url: https://charts.bitnami.com/bitnami
        release_namespace: monitoring
```
```
ansible --version
ansible 2.9.2
  config file = /home/keiran/Ansible/ansible.cfg
  configured module search path = ['/home/keiran/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python3.7/site-packages/ansible
  executable location = /usr/bin/ansible
  python version = 3.7.4 (default, Oct  4 2019, 06:57:26) [GCC 9.2.0]
```

T1loc commented:
Weird, it's working well now on my side too. :-/ 
So for me everything is good.

dstarner commented:
@acozine I see you approved/merged a PR elsewhere in this module's namespace, what are the next steps to see this get merged in? Are you a valid reviewer for this potential addition?

PPACI commented:
Does someone have an update on this ? If something is needed, I could manage to find some time to help on it.

dstarner commented:
@PPACI I *think* it just needs a review from the proper individuals, but Im not quite sure where (or who) those people are.

bkimbrough88 commented:
> @PPACI I _think_ it just needs a review from the proper individuals, but Im not quite sure where (or who) those people are.

Do you know if Ansible has community office hours or the like?

PPACI commented:
Sorry to bump again without any real reason, but we still don't have any official information for months now.
Is something still blocking the PR ?

gebi commented:
would be awesome to get this merged :)!

felixfontein commented on lib/ansible/modules/cloud/misc/helm_cli.py:
```diff
@@ -0,0 +1,595 @@
+      - absolute URL
+      - Required when release_state is set to present
+    required: false
+    type: str
```
Maybe use
```suggestion
    type: path
```
instead? Then Ansible will replace things like `~` IIRC.

felixfontein commented on lib/ansible/modules/cloud/misc/helm_cli.py:
```diff
@@ -0,0 +1,595 @@
+      - path to a packaged chart
+      - path to an unpacked chart directory
+      - absolute URL
+      - Required when release_state is set to present
```
```suggestion
      - Required when I(release_state) is set to C(present)
```
Use `I(...)` for option names, `C(...)` for option values.

felixfontein commented on lib/ansible/modules/cloud/misc/helm_cli.py:
```diff
@@ -0,0 +1,595 @@
+    type: str
+  chart_repo_url:
+    description:
+      - Chart repository url where to locate the requested chart
```
```suggestion
      - Chart repository URL where to locate the requested chart
```

felixfontein commented on lib/ansible/modules/cloud/misc/helm_cli.py:
```diff
@@ -0,0 +1,595 @@
+  chart_repo_username:
+    description:
+      - Chart repository username where to locate the requested chart
+      - Required if chart_repo_password is specified
```
```suggestion
      - Required if I(chart_repo_password) is specified
```

felixfontein commented on lib/ansible/modules/cloud/misc/helm_cli.py:
```diff
@@ -0,0 +1,595 @@
+  chart_repo_password:
+    description:
+      - Chart repository password where to locate the requested chart
+      - Required if chart_repo_username is specified
```
```suggestion
      - Required if I(chart_repo_username) is specified
```

felixfontein commented on lib/ansible/modules/cloud/misc/helm_cli.py:
```diff
@@ -0,0 +1,595 @@
+    type: str
+  update_repo_cache:
+    description:
+      - Run `helm repo update` before the operation. Can be run as part of the package installation or as a separate step.
```
```suggestion
      - Run C(helm repo update) before the operation. Can be run as part of the package installation or as a separate step.
```

felixfontein commented on lib/ansible/modules/cloud/misc/helm_cli.py:
```diff
@@ -0,0 +1,595 @@
+  update_repo_cache:
+    description:
+      - Run `helm repo update` before the operation. Can be run as part of the package installation or as a separate step.
+    default: "false"
```
```suggestion
    default: false
```
`"false"` is a string, not a boolean

felixfontein commented on lib/ansible/modules/cloud/misc/helm_cli.py:
```diff
@@ -0,0 +1,595 @@
+    type: bool
+  wait_timeout:
+    description:
+      - Timeout when wait option is enabled (helm2 is a number of seconds, helm3 is a duration
```
```suggestion
      - Timeout when wait option is enabled (helm2 is a number of seconds, helm3 is a duration)
```
Is it clear what this means to a user of the module?

felixfontein commented on lib/ansible/modules/cloud/misc/helm_cli.py:
```diff
@@ -0,0 +1,595 @@
+
+# Get Release state from deployed release
+def get_release_status(command, release_name, release_namespace):
+    list_command = command + " list --output=yaml "
```
I guess it is a bit late now, but in general, avoid composing commands as strings. Use a list instead with every part of the command in one element. That way, you won't have problems with escaping stuff or having spaces or other crazy things in filenames.

T1loc commented:
One thing I found, when you enable the update_repo_cache and because the helm 3 client is not adding default repository we get an error message : 

```
fatal: [localhost]: FAILED! => {
    "changed": false,
    "command": "/usr/local/bin/helm --kube-context k8s-dev repo update"
}

MSG:

Failure when executing Helm command. Exited 1.
stdout:
stderr: Error: no repositories found. You must add one before updating
```

```
helm repo update
Error: no repositories found. You must add one before updating
```

For my use, I can just set update_repo_cache to false, because I explicitly set my repository for each chart I want to install. 
However it would be great to handle this case, isn't ?

geerlingguy commented:
Just noting that it's been proposed that the `helm` module be maintained inside the Kubernetes collection (see https://github.com/ansible-collections/kubernetes/issues/28)—and I somewhat agree it might be a good place for it, since Helm is really only used in the context of Kubernetes.

The PR might not be a quick shift over to that collection's repository, but it might be worth doing, and then also marking the helm module in Ansible 2.9+ as deprecated, or something along those lines (I know there's also going to be the ability to mark a module as 'migrated' in 2.10).

cc @tima

LucasBoisserie commented:
Ansible core doesn't not accept any new module anymore
PR moved to https://github.com/ansible-collections/community.kubernetes and merged on https://github.com/ansible-collections/community.kubernetes/pull/61
